Add support for SOCKS5 private authentication methods (RFC#1928)

sunil-solace opened the pull request:
Motivation:
The SOCKS5 protocol, as defined by RFC 1928, reserves authentication method codes from 0x80 to 0xFE for private or experimental use. The current Socks5ProxyHandler in Netty only supports the standard methods: No Authentication (0x00) and GSS-API / Username/Password (0x01/0x02). This limitation prevents users from implementing and utilizing custom authentication schemes with SOCKS5 proxy servers that require private authentication methods.

This change addresses this gap by adding support for these private authentication methods, as requested in issue #15460.


Modification:
Introduced a new constructor in Socks5ProxyHandler that accepts a byte for the private authentication method and a byte[] for the authentication token/payload.
When this new constructor is used, the handler will offer the specified private method to the SOCKS5 proxy during the initial handshake.
If the proxy selects the private method, the handler proceeds with the private authentication sub-negotiation by sending the provided token.
Added corresponding tests in ProxyHandlerTest.java to validate the new functionality. This includes creating a test Socks5ProxyServer that supports a private authentication method and adding test cases for successful authentication, authentication failure (bad token), and connection rejection.

Result:
Fixes #15460.

This enhancement allows users to connect to SOCKS5 proxies that require custom, non-standard authentication, increasing the flexibility and completeness of Netty's proxy support.

sunil-solace commented:
@normanmaurer @trustin I'd like to request your review on this PR. This contribution implements SOCKS5 private authentication method support as discussed in issue #15460 (https://github.com/netty/netty/issues/15460).

normanmaurer commented on codec-socks/src/main/java/io/netty/handler/codec/socksx/v5/DefaultSocks5PrivateAuthRequest.java:
```diff
@@ -0,0 +1,67 @@
+/**
+ * The default {@link Socks5PrivateAuthRequest} implementation.
+ * <p>
+ * For custom private authentication protocols, this class can be extended to include
```
This class is final so it can't be extended... Maybe you want to say you can implement the interface ?

normanmaurer commented on codec-socks/src/main/java/io/netty/handler/codec/socksx/v5/DefaultSocks5PrivateAuthRequest.java:
```diff
@@ -0,0 +1,67 @@
+   *
+   * @param privateAuthToken the private authentication token
+   */
+  public DefaultSocks5PrivateAuthRequest(final byte[] privateAuthToken) {
```
We use 4 spaces and not 2... this is true for everywhere.

sunil-solace replied:
taken care throughout!

normanmaurer commented on codec-socks/src/main/java/io/netty/handler/codec/socksx/v5/DefaultSocks5PrivateAuthRequest.java:
```diff
@@ -0,0 +1,67 @@
+  public DefaultSocks5PrivateAuthRequest(final byte[] privateAuthToken) {
+    ObjectUtil.checkNotNull(privateAuthToken, "privateToken");
+
+    this.privateToken = privateAuthToken;
```
```suggestion
    this.privateToken = privateAuthToken.clone();
```

normanmaurer commented on codec-socks/src/main/java/io/netty/handler/codec/socksx/v5/DefaultSocks5PrivateAuthRequest.java:
```diff
@@ -0,0 +1,67 @@
+
+  @Override
+  public byte[] privateToken() {
+    return privateToken;
```
```suggestion
    return privateToken.clone();
```

normanmaurer commented on codec-socks/src/main/java/io/netty/handler/codec/socksx/v5/DefaultSocks5PrivateAuthResponse.java:
```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2023 The Netty Project
```
```suggestion
 * Copyright 2025 The Netty Project
```

normanmaurer commented on codec-socks/src/main/java/io/netty/handler/codec/socksx/v5/Socks5AuthMethod.java:
```diff
@@ -32,6 +32,18 @@ public class Socks5AuthMethod implements Comparable<Socks5AuthMethod> {
 
+    /**
+     * Returns whether the authentication method code is in the private methods range (0x80-0xFE)
+     * as defined by RFC 1928 section 3.
```
nit: consider using `<a href....`

normanmaurer commented on codec-socks/src/main/java/io/netty/handler/codec/socksx/v5/Socks5PrivateAuthRequestDecoder.java:
```diff
@@ -0,0 +1,93 @@
+ * On successful decode, this decoder will forward the received data to the next handler, so that
+ * other handler can remove or replace this decoder later.
+ */
+public class Socks5PrivateAuthRequestDecoder extends ReplayingDecoder<Socks5PrivateAuthRequestDecoder.State> {
```
Please extend and use `ByteToMessageDecoder` it's fare better in terms of performance

normanmaurer commented on codec-socks/src/main/java/io/netty/handler/codec/socksx/v5/Socks5PrivateAuthRequestDecoder.java:
```diff
@@ -0,0 +1,93 @@
+                    throw new DecoderException("unsupported subnegotiation version: " + version + " (expected: 1)");
+                }
+
+                int readableBytes = actualReadableBytes();
```
this does not sound correct... there needs to be some notion how much data is expected.

normanmaurer commented on codec-socks/src/main/java/io/netty/handler/codec/socksx/v5/Socks5PrivateAuthResponse.java:
```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright 2023 The Netty Project
```
```suggestion
 * Copyright 2025 The Netty Project
```

normanmaurer commented on codec-socks/src/main/java/io/netty/handler/codec/socksx/v5/Socks5PrivateAuthResponseDecoder.java:
```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright 2023 The Netty Project
```
```suggestion
 * Copyright 2025 The Netty Project
```

normanmaurer commented on codec-socks/src/main/java/io/netty/handler/codec/socksx/v5/Socks5PrivateAuthResponseDecoder.java:
```diff
@@ -0,0 +1,120 @@
+ * </ul>
+ * </p>
+ */
+public final class Socks5PrivateAuthResponseDecoder extends ReplayingDecoder<Socks5PrivateAuthResponseDecoder.State> {
```
same comment as above

normanmaurer commented on codec-socks/src/main/java/io/netty/handler/codec/socksx/v5/Socks5PrivateAuthStatus.java:
```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright 2023 The Netty Project
```
```suggestion
 * Copyright 2025 The Netty Project
```

normanmaurer commented on handler-proxy/src/main/java/io/netty/handler/proxy/Socks5ProxyHandler.java:
```diff
@@ -86,7 +117,14 @@ public String protocol() {
+        Socks5AuthMethod authMethod = socksAuthMethod();
+        if (Socks5AuthMethod.isPrivateMethod(authMethod.byteValue())) {
+            return AUTH_PRIVATE;
+        } else if (authMethod == Socks5AuthMethod.PASSWORD) {
```
```suggestion
        }
        if (authMethod == Socks5AuthMethod.PASSWORD) {
```

normanmaurer commented on handler-proxy/src/main/java/io/netty/handler/proxy/Socks5ProxyHandler.java:
```diff
@@ -86,7 +117,14 @@ public String protocol() {
+            return AUTH_PRIVATE;
+        } else if (authMethod == Socks5AuthMethod.PASSWORD) {
+            return AUTH_PASSWORD;
+        } else {
+            return AUTH_NONE;
+        }
```
```suggestion
        return AUTH_NONE;
```

sunil-solace commented:
> @sunil-solace did you sign our ICLA ? https://netty.io/s/icla

Yes @normanmaurer we have signed the CCLA and I've been designated as the primary contributor for our organization.

sunil-solace commented:
@normanmaurer  I've completed all the requested changes from your previous review feedback. 
I appreciate your time and expertise in reviewing this contribution. The PR is now ready for your review again when you have a moment.

yawkat commented:
Norman is currently on PTO, it may take a while.

sunil-solace commented:
@normanmaurer 
Hope you had a refreshing break! 
Appreciate the recent review. I've addressed all the comments. Could you please take another look when you have a chance?

normanmaurer commented on codec-socks/src/main/java/io/netty/handler/codec/socksx/v5/Socks5PrivateAuthStatus.java:
```diff
@@ -0,0 +1,96 @@
+
+    public Socks5PrivateAuthStatus(int byteValue, String name) {
+        this.name = ObjectUtil.checkNotNull(name, "name");
+        this.byteValue = (byte) byteValue;
```
ah ok ..

sunil-solace commented on codec-socks/src/main/java/io/netty/handler/codec/socksx/v5/Socks5PrivateAuthRequestDecoder.java:
```diff
@@ -0,0 +1,93 @@
+
+        decoded = true;
+
+        Socks5Message m = new DefaultSocks5PrivateAuthRequest(Unpooled.EMPTY_BUFFER.array());
```
wasn't aware of this utility, updated!

sunil-solace commented:
@normanmaurer 
When you have a moment, could you please verify that all review comments have been resolved? Appreciate your time!

normanmaurer commented on codec-socks/src/main/java/io/netty/handler/codec/socksx/v5/Socks5PrivateAuthStatus.java:
```diff
@@ -0,0 +1,96 @@
+    private final String name;
+    private String text;
+
+    public Socks5PrivateAuthStatus(int byteValue) {
```
Make this private as the static method should be used. 

```suggestion
    private Socks5PrivateAuthStatus(int byteValue) {
```
